wgpu-hal/src/dx12/device.rs

```diff
@@ -221,64 +222,147 @@ impl super::Device {
+                hlsl::ShaderModel::V5_1.to_str()
+            );
+            let mut error = native::Blob::null();
+
+            let hr = unsafe {
+                profiling::scope!("d3dcompiler::D3DCompile");
+                d3dcompiler::D3DCompile(
+                    source.as_ptr() as *const _,
+                    source.len(),
+                    source_name.as_ptr() as *const i8,
+                    ptr::null(),
+                    ptr::null_mut(),
+                    raw_ep.as_ptr(),
+                    full_stage.as_ptr() as *const i8,
+                    compile_flags,
+                    0,
+                    shader_data.mut_void() as *mut *mut _,
+                    error.mut_void() as *mut *mut _,
+                )
+            };
 
-        let (result, log_level) = match hr.into_result() {
-            Ok(()) => (Ok(shader_data), log::Level::Info),
-            Err(e) => {
-                let mut full_msg = format!("D3DCompile error ({})", e);
-                if !error.is_null() {
-                    use std::fmt::Write as _;
-                    let message = unsafe {
-                        slice::from_raw_parts(
-                            error.GetBufferPointer() as *const u8,
-                            error.GetBufferSize(),
-                        )
-                    };
-                    let _ = write!(full_msg, ": {}", String::from_utf8_lossy(message));
-                    unsafe {
-                        error.destroy();
+            match hr.into_result() {
+                Ok(()) => (Ok(super::ShaderDXIL::FXC(shader_data)), log::Level::Info),
+                Err(e) => {
+                    let mut full_msg = format!("D3DCompile error ({})", e);
+                    if !error.is_null() {
+                        use std::fmt::Write as _;
+                        let message = unsafe {
+                            slice::from_raw_parts(
+                                error.GetBufferPointer() as *const u8,
+                                error.GetBufferSize(),
+                            )
+                        };
+                        let _ = write!(full_msg, ": {}", String::from_utf8_lossy(message));
+                        unsafe {
+                            error.destroy();
+                        }
                     }
+                    (
+                        Err(crate::PipelineError::Linkage(stage_bit, full_msg)),
+                        log::Level::Warn,
+                    )
                 }
-                (
-                    Err(crate::PipelineError::Linkage(stage_bit, full_msg)),
-                    log::Level::Warn,
-                )
             }
         };
 
+        // hassle_rs doesn't check for `dxil.dll` if you only call `hassle_rs::compile_hlsl` but calling hassle_rs::Dxil::new will .
+        profiling::scope!("hassle_rs::compile_hlsl");
+        let (result, log_level) = match hassle_rs::Dxil::new(None).and(hassle_rs::compile_hlsl(
```

Comment:
You may want to make this `and_then` so that `hassle_rs::compile_hlsl` isn't called if `Dxil` fails.

Reply:
nice catch !